### PR TITLE
Add setting values from json file

### PR DESCRIPTION
**Pull request by seeemilyplay**
This adds a way to set values from a json file.

@zimbatm what do you reckon?

I don't quite understand the codebase fully, but something like this would be perfect for the automation I'm working on at the moment.

### CODE REVIEW

File: bin/stronghold-cli [hunk 1]
```diff
@@ -45,6 +45,18 @@ def intract(res, path, val)
   res
 end
 
+def intract_json(res, json)
+  res.update(json) do |key, old_value, new_value|
+    if old_value.class == Hash && !force
+      die(-12, "Won't overwrite hash in update of #{path}, would overwrite #{old_value}")
```
Comment (zimbatm) [inline bin/stronghold-cli]: I recommend using `old_value.inspect` to get a better output

File: bin/stronghold-cli [hunk 2]
```diff
@@ -45,6 +45,18 @@ def intract(res, path, val)
...
+    elsif new_value.class == Hash
+      die(-13, "Won't update #{path} with hash #{new_value}")
```
Comment (zimbatm) [inline bin/stronghold-cli]: Same

**Comment by zimbatm** [comment]
Nice ! I think that `--set-json` who reads from stdin would be a bit more versatile. We could even have a `--get-json` if you want.

What happens if the old data == the new data, does it still make a commit ?

**Comment by seeemilyplay** [comment]
@zimbatm, good point about when old data == new data, it did a new commit. I've changed this, and also added a --set-json and --get-json. The set-json uses an arg rather than the stdin, just cos it seemed easier (to me - I only got as far as half googling "ruby stdin"). You can always use "xargs -0" to pipe in a file with that. What do you reckon?

File: bin/stronghold-cli [hunk 3]
```diff
@@ -166,13 +180,29 @@ opts.each do |opt, arg|
...
+        data: new_data,
+        comment: "At level #{$path} set from file #{arg}"
+      })
+    end
   when '--get'
     mode_ify(opt)
     puts format(extract(valid_tree(), data_path(arg)))
+  when '--get-json'
+    mode_ify(opt)
+    puts format(client.head.tree.peculiar($path))
```
Comment (zimbatm) [inline bin/stronghold-cli]: Actually I forgot that `--get` already returns JSON. Is this one returning the non-materialized view ?

File: bin/stronghold-cli [hunk 4]
```diff
@@ -166,13 +180,29 @@ opts.each do |opt, arg|
...
     client.head.update({
         :path => $path,
-        :author => "command-line on #{Socket.gethostname}",
+        :author => "command-line on #{Socket.gethostname}, USER=#{ENV['USER']}",
         :data => intract(client.head.tree.peculiar($path), internal_path, to),
         :comment => comment || "At level #{$path} set #{arg}"
       })
+  when '--set-json'
+    mode_ify(opt)
+    json = JSON.parse(arg)
```
Comment (zimbatm) [inline bin/stronghold-cli]: you can use `JSON.load($stdin)` for the streaming version if you want

File: bin/stronghold-cli [hunk 5]
```diff
@@ -166,13 +180,29 @@ opts.each do |opt, arg|
...
+    old_data = client.head.tree.peculiar($path)
+    new_data = intract_json(old_data, json)
+    unless old_data == new_data
+      client.head.update({
```
Comment (zimbatm) [inline bin/stronghold-cli]: nit: the `{}` are superfluous when the last argument is a hash and generally avoided in ruby.

File: bin/stronghold-cli (hunk 3 above)
Comment (mdpye) [inline bin/stronghold-cli]: Yeah, suggest we call it `--get-peculiar` instead.

File: bin/stronghold-cli (hunk 4 above)
Comment (mdpye) [inline bin/stronghold-cli]: Would prefer that, taking a huge text blob as a command line argument is... unorthodox.

**Comment by seeemilyplay** [comment]
Ks, think I've got all the feedback incorporated now. Thanks for the reviews @zimbatm @mdpye. Thumbs up for merge?

**Comment by zimbatm** [comment]
:beers:

**Pull request merged**